mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMailboxMapper.java

```diff
@@ -256,30 +263,23 @@ public MailboxId rename(Mailbox mailbox) throws MailboxException {
     }
 
     @Override
-    public List<Mailbox> list() throws MailboxException {
-       File maildirRoot = maildirStore.getMaildirRoot();
-       List<Mailbox> mailboxList = new ArrayList<>();
-
-       if (maildirStore.getMaildirLocation().endsWith("/" + MaildirStore.PATH_DOMAIN + "/" + MaildirStore.PATH_USER)) {
-           File[] domains = maildirRoot.listFiles();
-           for (File domain : domains) {
-               File[] users = domain.listFiles();
-               visitUsersForMailboxList(domain, users, mailboxList);
-           }
-           return mailboxList;
-       }
-
-        File[] users = maildirRoot.listFiles();
-        visitUsersForMailboxList(null, users, mailboxList);
-        return mailboxList;
+    public Flux<Mailbox> list() {
+        File maildirRoot = maildirStore.getMaildirRoot();
+        return Mono.fromCallable(maildirStore::getMaildirLocation)
+            .filter(dir -> dir.endsWith("/" + MaildirStore.PATH_DOMAIN + "/" + MaildirStore.PATH_USER))
+            .map(ignored -> Arrays.stream(maildirRoot.listFiles())
+                .flatMap(Throwing.<File, Stream<Mailbox>>function(domain -> visitUsersForMailboxList(domain, domain.listFiles()).stream()).sneakyThrow()))
+            .switchIfEmpty(Mono.fromCallable(() -> visitUsersForMailboxList(null, maildirRoot.listFiles()).stream()))
+            .flatMapIterable(mailboxes -> mailboxes.collect(Guavate.toImmutableList()));
     }
 
     @Override
     public void endRequest() {
 
     }
 
-    private void visitUsersForMailboxList(File domain, File[] users, List<Mailbox> mailboxList) throws MailboxException {
+    private List<Mailbox> visitUsersForMailboxList(File domain, File[] users) throws MailboxException {
+        ImmutableList.Builder<Mailbox> mailboxList = ImmutableList.builder();
         String userName = null;
         
         for (File user: users) {
```

Comment:
May be a short if here would be nice?

Reply:
Sorry?

Reply:
I mean the if else block inside the for loop. I can't seem to comment on it.

Reply:
Sorry but maildir code should be considered legacy.

Changes in there should be considered legacy.

If you cannot comment on it it is for a good reason: it is not part of the changeset.

I *could* enhence the code here, but 1. we won't know where to stop, 2. it is not really related with the work I am doing here.